### PR TITLE
Feat fraud detection

### PR DESCRIPTION
**Pull request by PedroCaldeira**
This PR implements a fully fledged end to end prototype of the fraud.
Creating a new fraud service implemented in python and flask.

### CODE REVIEW

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    private String fraudServiceURL;
+
+    @Autowired
+    private QuizService quizService;
+    
+    @Autowired
+    private AnswerService answerService;
+
+    @GetMapping("/fraud/execution/{executionId}")
+    @PreAuthorize("hasRole('ROLE_TEACHER') and hasPermission(#quizId, 'QUIZ.ACCESS')")
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java:41]: {HIGH} #executionId, 'EXECUTION.ACCESS'

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    public List<QuizFraudScoreDto> getCourseExecutionFraudScores(@PathVariable Integer executionId) {
+        return getFraudScoresFromURI("/execution/" + executionId.toString());
+    }
+
+    @GetMapping("/fraud/execution/{executionId}/user/{userId}")
+    @PreAuthorize("hasRole('ROLE_TEACHER') and hasPermission(#quizId, 'QUIZ.ACCESS')")
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java:47]: {HIGH} #executionId, 'EXECUTION.ACCESS'

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    public List<QuizFraudScoreDto> getUserFraudScores(@PathVariable Integer executionId, @PathVariable Integer userId) {
+        return getFraudScoresFromURI("/execution/" + executionId.toString() + "/user/" + userId.toString());
+    }
+
+    @GetMapping("/fraud/quiz/{quizId}")
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java:52]: to give more semantics to the name, fraud/time/quiz/{quizId}

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    @PreAuthorize("hasRole('ROLE_TEACHER') and hasPermission(#quizId, 'QUIZ.ACCESS')")
+    public List<QuizFraudScoreDto> getQuizFraudScores(@PathVariable Integer quizId) {
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java:54]: getQuizFraudTimeScores

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        answerService.writeQuizAnswers(quizId);
+        validateQuizId(quizId);
+        return getFraudScoresFromURI("/quiz/" + quizId.toString());
+    }
+
+    @GetMapping("/fraud/graph/quiz/{quizId}")
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java:60]: /fraud/communication/quiz/{quizId}

and apply consistently to the other cases

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/dto/QuizFraudGraphScoreDto.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package pt.ulisboa.tecnico.socialsoftware.tutor.fraud.dto;
+
+import java.io.Serializable;
+import java.util.List;
+
+public class QuizFraudGraphScoreDto implements Serializable{
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/dto/QuizFraudGraphScoreDto.java:6]: QuizFraudCommunicationScoreDto

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/dto/QuizFraudScoreDto.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package pt.ulisboa.tecnico.socialsoftware.tutor.fraud.dto;
+
+import java.io.Serializable;
+
+public class QuizFraudScoreDto implements Serializable{
```
Comment (ritosilva) [inline backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/dto/QuizFraudScoreDto.java:5]: QuizFraudTimeScoreDto

File: fraud-service/app.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+
+
+# Does not override env vars, so docker-compose shall set env vars if run. Otherwise, sets env vars from .env file
+load_dotenv()
+app = flask.Flask(__name__)
+
+
+@app.route('/quiz/<quizId>')
+def quizFraudScores(quizId):
```
Comment (ritosilva) [inline fraud-service/app.py:22]: include time

File: fraud-service/app.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+
+
+
+# Does not override env vars, so docker-compose shall set env vars if run. Otherwise, sets env vars from .env file
+load_dotenv()
+app = flask.Flask(__name__)
+
+
+@app.route('/quiz/<quizId>')
```
Comment (ritosilva) [inline fraud-service/app.py:21]: include time

File: fraud-service/app.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        quiz, quiz_stats.QUIZ_SCORERS["Diff. Mean (Question ID)"])
+    dbConnector.close()
+    
+    print("HI")
+    result = [{"username": k, "score": v}
+              for k, v in statistic.statistic.items()]
+    return flask.jsonify(result)
+
+
+@app.route('/graph/quiz/<quizId>')
```
Comment (ritosilva) [inline fraud-service/app.py:35]: replace graph by communication

and change other consistently

File: fraud-service/fraud_graph/create_dataset2.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
+import pandas as pd
+import queries_database2
+import create_query2
```
Comment (ritosilva) [inline fraud-service/fraud_graph/create_dataset2.py:3]: meaningful name

File: fraud-service/fraud_graph/create_query2.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
+def def_query(quiz_id = "*", course = "*", semester = "*", type_quiz = "*"):
```
Comment (ritosilva) [inline fraud-service/fraud_graph/create_query2.py:1]: meaningful name

File: fraud-service/fraud_graph/queries_database2.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
+import psycopg2
+import pandas.io.sql as psql
+import create_query2
+import os
+class Database(object):
+
+    def __init__(self):
+        
+        #A: connection details
```
Comment (ritosilva) [inline fraud-service/fraud_graph/queries_database2.py:9]: are the values hardcoded here, or the second argument is a default value?

File: frontend/src/services/RemoteServices.ts [hunk 14]
```diff
@@ -722,6 +723,32 @@ export default class RemoteServices {
       });
   }
 
+  static async getQuizFraudScores(quizId: number): Promise<QuizFraudScores> {
+    return httpClient
+      .get(`/fraud/quiz/${quizId}`)
```
Comment (ritosilva) [inline frontend/src/services/RemoteServices.ts:728]: time or times

File: frontend/src/services/RemoteServices.ts [hunk 15]
```diff
@@ -722,6 +723,32 @@ export default class RemoteServices {
...
+      })
+      .catch(async (error) => {
+        throw Error(await this.errorMessage(error));
+      });
+  }
+  static async getQuizGraphFraudScores(
+    quizId: number
+  ): Promise<QuizFraudScores[]> {
+    return httpClient
+      .get(`/fraud/graph/quiz/${quizId}`)
```
Comment (ritosilva) [inline frontend/src/services/RemoteServices.ts:740]: communication instead of graph

File: frontend/src/views/teacher/quizzes/QuizList.vue [hunk 16]
```diff
@@ -187,6 +198,15 @@
         />
       </v-card>
     </v-dialog>
+
+    <show-quiz-fraud-scores-dialog
+      v-if="quiz && quizFraudScores"
+      v-model="quizFraudScoresDialog"
+      :quiz="quiz"
+      :quizFraudScores="quizFraudScores"
```
Comment (ritosilva) [inline frontend/src/views/teacher/quizzes/QuizList.vue:206]: check the names, and use consistently, times and communication

File: frontend/src/views/teacher/quizzes/QuizList.vue [hunk 17]
```diff
@@ -295,6 +322,20 @@ export default class QuizList extends Vue {
     await this.$store.dispatch('clearLoading');
   }
 
+  async showQuizFraudScores(quiz: Quiz) {
+    await this.$store.dispatch('loading');
+    try {
```
Comment (ritosilva) [inline frontend/src/views/teacher/quizzes/QuizList.vue:327]: if this call is inside the dialog this component would be smaller and the call would be local to where the data is presented. would it be possible?

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java (hunk 4 above)
Comment (ritosilva) [reply backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/FraudController.java:54]: apply the some idea do the others

**Review by ritosilva — CHANGES_REQUESTED** [review]
[reviewpad-sig]: <>
[View on **Reviewpad**](https://beta.reviewpad.com/review/github.com/socialsoftware/quizzes-tutor/pull/349#review-304cbbe8-b130-4b12-ac65-eb1e3bfec4d8)

File: backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/dto/QuizFraudScoreDto.java (hunk 7 above)
Comment (PedroCaldeira) [reply backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/fraud/dto/QuizFraudScoreDto.java:5]: This is the Building DTO Object for a fraud entry, consisting of a user and a score. I'll add another class for the specialization of a QuizFraudTimeScoreDto and rename this to UserFraudScoreDto

**Pull request merged**